zephyr: Use Kconfig to configure MicroPython options.

MaureenHelm opened the pull request:
The zephyr port uses Kconfig to enable/disable and configure zephyr
kernel features at build time, such as peripheral drivers, networking,
and usb. Extend the Kconfig infrastructure to include MicroPython
options in mpconfigport.h.

This allows us to combine mpconfigport.h and mpconfigport_minimal.h into
one file, and to configure zephyr and MicroPython options together in
one place. Before this change, it was difficult to enable the
MicroPython machine.i2c class for only boards that have the zephyr i2c
driver enabled. After this change, we can now enable the class in the
same place we enable the driver. For example, in
ports/zephyr/boards/frdm_k64f.conf:

  CONFIG_I2C=y
  CONFIG_MICROPY_PY_MACHINE_I2C=y

This change also allows us to configure MicroPython options with menuconfig.

Kconfig options are currently added only for MicroPython options that
were explicitly configured in zephyr's mpconfigport.h or
mpconfigport_minimal.h; others can be added in the future for further
refinement. Most options are implicitly default n to optimize for code
size. Those that are default y were previously enabled in both
mpconfigport.h and mpconfigport_minimal.h.

To verify that the MicroPython configuration hasn't changed after moving
options to Kconfig, we can compare flash sizes before and after this
commit, using the zephyr sdk 0.11.3 toolchain and zephyr v2.3.0:

|                                |Before     |After|
|---|---|---|
|`$ ./make-minimal BOARD=frdm_k64f`  |122300 B   |122328 B|
|`$ make BOARD=frdm_k64f`            |239652 B   |239652 B|

The small increase in the minimal configuration is due to the fact that
it now includes the builtin open object.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

`$ make BOARD=frdm_k64f menuconfig`

![image](https://user-images.githubusercontent.com/6172539/87998437-8c517600-cabd-11ea-83a9-ea5829e22599.png)
![image](https://user-images.githubusercontent.com/6172539/87998478-abe89e80-cabd-11ea-96f9-a604e22fe02f.png)
![image](https://user-images.githubusercontent.com/6172539/87998483-af7c2580-cabd-11ea-93c7-f76dbd3df810.png)

MaureenHelm committed:
zephyr: Fix modusocket build when MICROPY_PY_USOCKET=0.

Fixes the zephyr usocket module to build successfully when
MICROPY_PY_USOCKET is defined but set to zero.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
zephyr: Support building without machine module.

In preparation for combining zephyr's mpconfigport.h and
mpconfigport_minimal.h files, support building without the machine
module.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
zephyr: Use Kconfig to configure MicroPython options.

The zephyr port uses Kconfig to enable/disable and configure zephyr
kernel features at build time, such as peripheral drivers, networking,
and usb. Extend the Kconfig infrastructure to include MicroPython
options in mpconfigport.h.

This allows us to combine mpconfigport.h and mpconfigport_minimal.h into
one file, and to configure zephyr and MicroPython options together in
one place. Before this change, it was difficult to do things like enable
the MicroPython machine.i2c class for only boards that have the zephyr
i2c driver enabled. After this change, we can now enable the class in
the same place we enable the driver. For example, in
ports/zephyr/boards/frdm_k64f.conf:

  CONFIG_I2C=y
  CONFIG_MICROPY_PY_MACHINE_I2C=y

This change also allows us to configure MicroPython options with
menuconfig.

Kconfig options are currently added only for MicroPython options that
were explicitly configured in zephyr's mpconfigport.h or
mpconfigport_minimal.h; others can be added in the future for further
refinement. Most options are implicitly default n to optimize for code
size. Those that are default y were previously enabled in both
mpconfigport.h and mpconfigport_minimal.h.

To verify that the MicroPython configuration hasn't changed after moving
options to Kconfig, we can compare flash sizes before and after this
commit, using the zephyr sdk 0.12.2 toolchain, zephyr v2.5.0, and the
frdm_k64f board:

                  Before      After
Minimal config    114524 B    114548 B
Normal config     236644 B    236644 B

The small increase in the minimal configuration is due to the fact that
it now includes the builtin open object.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
zephyr: Disable MICROPY_PY_BUILTINS_STR_COUNT in minimal build.

MICROPY_PY_BUILTINS_STR_COUNT is disabled in the normal build; it should
also be disabled in the minimal build.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

dlech commented:
Have you seen #5444? Any insights to offer there after having implemented this PR?

MaureenHelm commented:
> Have you seen #5444? Any insights to offer there after having implemented this PR?

My motivation is perhaps slightly different and more narrowly focused, but not incompatible with yours. For the zephyr port, it is confusing and error-prone to configure things in two places (Kconfig and mpconfigport.h). I agree with most of your arguments in #5444 (except the Kbuild part) and would be in favor of using Kconfig throughout. Zephyr has been using Kconfig for 4+ years and currently supports over 200 board configurations.

I don't have a good sense of how the MicroPython community would feel about such an intrusive change though. Given that the zephyr port already uses Kconfig, maybe we use this port as a trial?

One thing to consider is potential Kconfig symbol name conflicts between the two projects, such as CONFIG_BOARD.

You should also look at #6006, which takes advantage of Kconfig overlays.

MaureenHelm commented:
@dpgeorge any feedback?

MaureenHelm commented:
Rebased to latest master and resolved merge conflicts.

@dpgeorge ping

MaureenHelm commented:
Rebased to latest master and updated the before/after flash sizes to verify that we're still not inadvertently changing the MicroPython configuration.

@dpgeorge It looks like the esp32 port also uses Kconfig. Might it benefit from this change too?
